### PR TITLE
Shortest path implementation

### PR DESCRIPTION
**Pull request by adfeel220**
# Shortest Path

## Updates

- The shortest path algorithm is supported with directed and undirected graphs
    - The return type is a vector of vertices as its path
- Current implementation only works when the source and destination are specified.

## Testing

- The test script creates a random graph, selecting source and destination vertices, and compares the shortest path obtained by `Graphs.dijkstra_shortest_path` and this result.

### CODE REVIEW

**Comment by gdalle** [comment]
Just merged `main` into your branch, that way we should see the docs at https://juliagraphs.org/GraphsOptim.jl/previews/PR17

EDIT: crap, we can't because you [pushed from a fork](https://github.com/JuliaGraphs/GraphsOptim.jl/actions/runs/6379604190/job/17312384208?pr=17#step:5:330)

File: docs/src/algorithms.md [hunk 1]
```diff
@@ -42,6 +42,50 @@ where the incoming flow $f^-(v)$ and outgoing flow $f^+(v)$ are defined as
...
+
+The variable $x(u,v)$ is an edge selection variable. $x(u,v) = 1$ means that edge $(u, v)$ is selected in the path and $x(u, v)=  0$ otherwise.
+
+The linear program is constrained by
+
+1. 0 to 1 selection for all edges:
+```math
+0 \leq x(u, v) \leq 1
+```
+
```
Comment (gdalle) [inline docs/src/algorithms.md]: this screws up the numbering

File: src/GraphsOptim.jl [hunk 2]
```diff
@@ -1,11 +1,11 @@
 """
-    GraphsOptim
+	GraphsOptim
 
 A package for graph optimization algorithms that rely on mathematical programming.
 """
 module GraphsOptim
 
-using Graphs: AbstractGraph, is_directed
+using Graphs: AbstractGraph, is_directed, DiGraph
```
Comment (gdalle) [inline src/GraphsOptim.jl]: why is this needed?

File: src/shortest_path.jl [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    model::Model,
+    g::AbstractGraph,
+    source::Int,
+    target::Int,
+    edge_cost::AbstractMatrix;
+    var_name,
+)
+    dg = g
+    if !is_directed(g)
+        dg = DiGraph(g)
```
Comment (gdalle) [inline src/shortest_path.jl]: not necessary, but test with undirected graphs

File: src/shortest_path.jl [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    end
+    edge_tuples = [(src(ed), dst(ed)) for ed in edges(dg)]
+
+    # 1 if an edge if traversed and 0 otherwise
+    traversed = @variable(model, 0 <= x[e=edge_tuples] <= 1; integer=true)
```
Comment (gdalle) [inline src/shortest_path.jl]: give the option integer or not

File: src/shortest_path.jl [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    edge_selection = value.(model[:path])
+    valid_edges = [p for p in first(axes(edge_selection)) if edge_selection[p] > 0]
+
+    # Parse valid path indices to actual path (sequence of traversed vertices)
+    path = ones(typeof(source), length(valid_edges) + 1)
+    path[begin] = source
+
+    for step in 1:length(valid_edges)
+        prev = path[step]
+        edge_index = findfirst(collect(edge[begin] == prev for edge in valid_edges))
```
Comment (gdalle) [inline src/shortest_path.jl]: collect not needed, findfirst used strangely
```julia
findfirst(edge -> edge[begin] == prev, valid_edges)
```

Comment (gdalle) [reply src/shortest_path.jl]: other option: select the edge with the highest weight going out of prev

File: test/shortest_path.jl [hunk 6]
```diff
@@ -0,0 +1,32 @@
+using Graphs
+using GraphsOptim
+using Test
+
+num_vertices = 20
+
+function test_shortest_path(g)
+    edge_cost = rand(1:100, num_vertices, num_vertices) .* adjacency_matrix(g)
```
Comment (gdalle) [inline test/shortest_path.jl]: better use fixed graphs for which we know the answer

File: test/shortest_path.jl [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        parent_vx = dijkstra_parents[dijkstra_ans[begin]]
+        pushfirst!(dijkstra_ans, parent_vx)
+        if parent_vx == source
+            break
+        end
+    end
+
+    mp_ans = GraphsOptim.shortest_path(g, source, target, edge_cost)
+
+    @test all(dijkstra_ans .== mp_ans)
```
Comment (gdalle) [inline test/shortest_path.jl]: test with Graphs.jl is okay for now, but networkx better in the future

**Review by gdalle — CHANGES_REQUESTED** [review]

File: src/GraphsOptim.jl (hunk 2 above)
Comment (adfeel220) [reply src/GraphsOptim.jl]: Removed, use the same algorithm for undirected graphs.

File: docs/src/algorithms.md (hunk 1 above)
Comment (adfeel220) [reply docs/src/algorithms.md]: Solved

File: src/shortest_path.jl (hunk 3 above)
Comment (adfeel220) [reply src/shortest_path.jl]: Removed and tested, it should be fine without it

File: src/shortest_path.jl (hunk 4 above)
Comment (adfeel220) [reply src/shortest_path.jl]: Integer option added

File: src/shortest_path.jl (hunk 5 above)
Comment (adfeel220) [reply src/shortest_path.jl]: Now use `first` function on an iterator

File: test/shortest_path.jl (hunk 6 above)
Comment (adfeel220) [reply test/shortest_path.jl]: Randomness removed and now uses shortest path example on Wikipedia for testing

**Review by gdalle — APPROVED** [review]
